src/notifier.rs

```diff
@@ -0,0 +1,55 @@
+use crate::constants::{MAX_NUMBER_OF_TRIES, SECS_TO_SLEEP_AFTER_REQUEST_ERROR};
+use serde_json::json;
+use std::error::Error;
+use std::thread::sleep;
+use std::time::Duration;
+
+pub struct TelegramNotifier {
+    token: String,
+    chat_id: String,
+}
+
+impl TelegramNotifier {
+    pub fn new(token: String, chat_id: String) -> TelegramNotifier {
+        TelegramNotifier { token, chat_id }
+    }
+
+    pub fn send_message(&self, message: &str) -> Result<(), Box<dyn Error>> {
+        tracing::info!("sending message to Telegram about new ads");
+        let client = reqwest::blocking::Client::new();
+
+        let mut number_of_tries = MAX_NUMBER_OF_TRIES;
+
+        loop {
+            let result = client
+                .post(format!(
+                    "https://api.telegram.org/bot{token}/sendMessage",
+                    token = &self.token
+                ))
+                .json(&json!({
+                    "text": message,
+                    "chat_id": self.chat_id
+                }))
+                .send();
+
+            match result {
+                Ok(_) => {
+                    tracing::info!("message is successfully sent to Telegram");
+                    break;
+                }
+                Err(err) => {
+                    tracing::error!("failed to send message to Telegram: {:?}", err);
+                    number_of_tries -= 1;
+                    tracing::info!("number of tries left: {}", number_of_tries);
+                    if number_of_tries == 0 {
+                        tracing::error!("number of tries exceeded");
+                        return Err(Box::new(err));
+                    }
+                    sleep(Duration::from_secs(SECS_TO_SLEEP_AFTER_REQUEST_ERROR));
```

Comment:
Don't you block the whole (and the single one) main thread just to wait?

Reply:
Yes, it's ok for the current bot version, because all operations are synchronous by the program logic: we first make a request to lun.ua, then check if conditions are met, then send tg notification, then go request lun.ua again.